lib/local-locks.js

```diff
@@ -0,0 +1,72 @@
+/**
+ * Implementation of a local JavaScript lock.
+ * Waiting threads (?) are placed in a queue and locks are obtained in the order
+ * that lock() is called.
+ */
+
+class LocalLock {
+    constructor() {
+        this.lock = false;
+        this.waiting = [];
+    }
+
+    /**
+     * Tries to acquire the lock.  If it is free this will return immediately.
```

Comment:
Would ideally like another set of eyes on this locking stuff to make sure something isn't (blatantly) wrong here